### PR TITLE
feat(goal_planner): reject candidate path whose start pose direction is not aligned with ego path

### PR DESCRIPTION
**Pull request by soblin**
## Description

In the `hasEnoughDistance`, use the "long_tail_reference_path" for checking if ego passed the start pose of pull_over_path candidate

image

## Related links

https://tier4.atlassian.net/browse/RT1-5675

https://evaluation.tier4.jp/evaluation/reports/14c05a40-8428-5835-8a08-bf923393f498?project_id=prd_jt

## Tests performed

https://github.com/autowarefoundation/autoware.universe/assets/28677420/8e896e62-f144-4d60-bf41-5c805c1ec062

https://evaluation.tier4.jp/evaluation/reports/e842342f-d2dd-59e3-897a-1f66799b0248?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [X] I've confirmed the [contribution guidelines].
- [X] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [X] The PR follows the [pull request guidelines].
- [X] The PR has been properly tested.
- [X] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [X] There are no open discussions or they are tracked via tickets.
- [X] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 1]
```diff
@@ -848,6 +848,25 @@ std::optional<std::pair<PullOverPath, GoalCandidate>> GoalPlannerModule::selectP
...
+    if (prev_path_back_to_goal_dist > backward_length) {
+      return prev_module_output_path;
+    }
+    // get road lanes which is at least backward_length[m] behind the goal
+    const auto road_lanes = utils::getExtendedCurrentLanesFromPath(
+      prev_module_output_path, planner_data_, backward_length, 0.0,
+      /*forward_only_in_route*/ false);
+    const auto goal_pose_length = lanelet::utils::getArcCoordinates(road_lanes, goal_pose).length;
+    return planner_data_->route_handler->getCenterLinePath(
+      road_lanes, std::max(0.0, goal_pose_length - backward_length), goal_pose_length);
```
Comment (kosuke55) [inline planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp:868]: for the case where modified goal is after original goal,
goal_pose_length + parameters_->forward_goal_search_length ?

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 2]
```diff
@@ -848,6 +848,25 @@ std::optional<std::pair<PullOverPath, GoalCandidate>> GoalPlannerModule::selectP
   const std::vector<PullOverPath> & pull_over_path_candidates,
   const GoalCandidates & goal_candidates, const double collision_check_margin) const
 {
+  const auto & goal_pose = planner_data_->route_handler->getOriginalGoalPose();
+  const double backward_length =
+    parameters_->backward_goal_search_length + parameters_->decide_path_distance;
+  const auto & prev_module_output_path = getPreviousModuleOutput().path;
+  const double prev_path_back_to_goal_dist = calcSignedArcLength(
+    prev_module_output_path.points, prev_module_output_path.points.back().point.pose.position,
+    goal_pose.position);
```
Comment (kosuke55) [inline planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp:857]: prev_module_output_path.points.front?

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 3]
```diff
@@ -1721,6 +1728,23 @@ bool GoalPlannerModule::hasEnoughDistance(const PullOverPath & pull_over_path) c
   // so need enough distance to restart.
   // distance to restart should be less than decide_path_distance.
   // otherwise, the goal would change immediately after departure.
+
+  // reject outdated shift path on curved roads whose start pose direction is far from ego yaw
```
[the remark this answers is not in the document]
Comment (soblin) [reply planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp]: fixed in f7a10caf9b2fdc9118104eb9d548ff0e56fec1af

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp (hunk 2 above)
Comment (soblin) [reply planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp]: fixed in f7a10caf9b2fdc9118104eb9d548ff0e56fec1af

File: planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp (hunk 1 above)
Comment (soblin) [reply planning/behavior_path_goal_planner_module/src/goal_planner_module.cpp]: fixed in f7a10caf9b2fdc9118104eb9d548ff0e56fec1af

**Review by kosuke55 — APPROVED** [review]
LGTM!

**Pull request merged**